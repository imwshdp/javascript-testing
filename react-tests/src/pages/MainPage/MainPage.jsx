import { useEffect, useState } from 'react';
import Users from '../../components/Users';

const MainPage = () => {
	const [data, setData] = useState(null);
	const [toggle, setToggle] = useState(false);
	const [inputValue, setInputValue] = useState('');

	const onToggleClick = () => setToggle(previousToggle => !previousToggle);

	useEffect(() => {
		const timer = setTimeout(() => {
			setData({});
		}, 100);

		return () => clearTimeout(timer);
	}, []);

	return (
		<section className='content'>
			<h1 data-testid='main-page'>Main Page</h1>

			<div>
				<h2>Example with text</h2>
				<span>Hello, world!</span>
			</div>

			<hr />

			<div>
				<h2>Example with toggle</h2>

				<button data-testid='toggle-switcher' onClick={onToggleClick}>
					Click me!
				</button>
				{toggle && <div data-testid='toggle-element'>Toggled!</div>}
			</div>

			<hr />

			<div>
				<h2>Example with input</h2>

				<input
					type='text'
					placeholder='input value'
					value={inputValue}
					onChange={event => setInputValue(event.target.value)}
				/>
				<div>
					<b data-testid='value-element'>{inputValue}</b>
				</div>
			</div>

			<hr />

			<div>
				<h2>Example with async actions</h2>

				{data && <div style={{ color: 'red' }}>Data is received</div>}
			</div>

			<hr />

			<div>
				<h2>Example with async data fetching</h2>

				<Users />
			</div>
		</section>
	);
};

export default MainPage;
