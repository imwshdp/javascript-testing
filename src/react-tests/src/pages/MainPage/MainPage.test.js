/* eslint-disable testing-library/no-debugging-utils */
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MainPage from './MainPage';

describe('React testing', () => {
	test('Examples with Get methods', () => {
		render(<MainPage />);

		const helloWorldElement = screen.getByText(/Hello, world!/i);
		expect(helloWorldElement).toBeInTheDocument();

		const buttonElement = screen.getByRole('button');
		expect(buttonElement).toBeInTheDocument();

		const inputElement = screen.getByPlaceholderText('input value');
		expect(inputElement).toBeInTheDocument();
		expect(inputElement).toMatchSnapshot();
	});

	test('Examples with Query methods', () => {
		render(<MainPage />);

		const nullableHelloWorldElement = screen.queryByText(/Hello, world! 2/i);
		expect(nullableHelloWorldElement).toBeNull();
	});

	test('Examples with Find methods', async () => {
		render(<MainPage />);

		screen.debug();

		const dataElement = await screen.findByText(/Data is received/i);
		expect(dataElement).toBeInTheDocument();
		expect(dataElement).toHaveStyle({
			color: 'red'
		});

		screen.debug();
	});

	test('Examples with React click event', () => {
		render(<MainPage />);

		const toggleSwitcher = screen.getByTestId('toggle-switcher');

		expect(screen.queryByTestId('toggle-element')).toBeNull();

		fireEvent.click(toggleSwitcher);
		expect(screen.getByTestId('toggle-element')).toBeInTheDocument();

		fireEvent.click(toggleSwitcher);
		expect(screen.queryByTestId('toggle-element')).toBeNull();
	});

	test('Examples with React input event', () => {
		render(<MainPage />);

		const inputElement = screen.getByPlaceholderText('input value');
		expect(screen.queryByTestId('value-element')).toContainHTML('');

		// Example with fireEvent
		// fireEvent.input(inputElement, {
		// 	target: {
		// 		value: '123'
		// 	}
		// });

		// Example with userEvent
		userEvent.type(inputElement, '123');

		expect(screen.queryByTestId('value-element')).toContainHTML('123');
	});
});
