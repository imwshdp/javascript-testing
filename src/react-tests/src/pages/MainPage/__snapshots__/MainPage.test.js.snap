// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React testing Examples with Get methods 1`] = `
<input
  placeholder="input value"
  type="text"
  value=""
/>
`;
